### PR TITLE
Feature: per-request fuel limit

### PR DESCRIPTION
**Pull request by cap10morgan**
Closes https://github.com/fluree/core/issues/33

This allows adding a `:fuel N` opt to transactions and queries and the fuel tracker will then ensure that your request stays below that fuel amount. If it does not, the request is short-circuited and an exception is returned.

I moved the existing fuel tests to the top-level `f.d.json-ld.api-test` ns b/c they were already testing fns in the `f.d.json-ld.api` ns. And then I added tests for this new per-req fuel limit. I also added CLJS fuel tests as well.

### CODE REVIEW

File: src/fluree/db/json_ld/transact.cljc [hunk 1]
```diff
@@ -535,25 +535,32 @@
...
+           flakes-ch     (if (q-parse/update? tx)
+                           (modify db fuel-tracker tx tx-state)
+                           (insert db fuel-tracker tx tx-state))
+           fuel-error-ch (:error-ch fuel-tracker)
+           flakes-chans  (remove nil? [fuel-error-ch flakes-ch])
+           [flakes]      (alts! flakes-chans :priority true)
+           _             (when (util/exception? flakes)
+                           (throw flakes))
+           _             (log/trace "stage flakes:" flakes)
+           final-chans   (remove nil? [fuel-error-ch
```
Comment (dpetran) [inline src/fluree/db/json_ld/transact.cljc:552]: I'm not sure I understand what's going on here - why do we still need the fuel-error-ch? Does flakes->final-db invoke the fuel tracker some more?

File: src/fluree/db/query/fql/parse.cljc [hunk 2]
```diff
@@ -488,6 +488,12 @@
     (assoc q :having (eval/compile code))
     q))
 
+(defn parse-fuel
```
[the remark this answers is not in the document]
Comment (zonotope) [reply src/fluree/db/query/fql/parse.cljc:491]: yeah, it seems like my pedantic QUERIES AND TRANSACTIONS MUST REMAIN SEPARATE arguments do have their negative implications.

**Review by zonotope — APPROVED** [review]
lgtm after @dpetran's comments are addressed.

File: src/fluree/db/api/query.cljc [hunk 3]
```diff
@@ -109,23 +109,28 @@
...
                      (<? (perm/wrap-policy db policy-opts))
                      db)
           db**     (-> (if t
                          (<? (time-travel/as-of db* t))
                          db*)
                        (assoc-in [:policy :cache] (atom {})))
-          query*   (-> query
+          query**  (-> query*
                        (update :opts assoc :issuer did)
                        (update :opts dissoc :meta)
```
Comment (dpetran) [inline src/fluree/db/api/query.cljc:124]: This is just a stylistic thing, so however the spirit leads, but you can put all the `dissoc` keys in one update.

File: src/fluree/db/query/exec/where.cljc [hunk 4]
```diff
@@ -218,6 +218,8 @@
                    start-flake (flake/create s* p o** o-dt* nil nil util/min-integer)
                    end-flake   (flake/create s* p o** o-dt* nil nil util/max-integer)
                    track-fuel  (when fuel-tracker
+                                 (take! (:error-ch fuel-tracker)
```
[the remark this answers is not in the document]
Comment (cap10morgan) [reply src/fluree/db/query/exec/where.cljc:221]: I'm not sure I follow. There is only one error channel per fuel tracker. And only one fuel tracker per request. The only place that seemed potentially different would be in the modify code path which uses queries and transactions. But I just checked and that code path already uses a single shared fuel-tracker. Let me know if I'm missing what you're pointing out, though.

**Comment by cap10morgan** [comment]
@dpetran & @zonotope let me know what you think of https://github.com/fluree/db/pull/586/commits/e170b706b02f64b7e9be84d065084441b554f3a2

File: src/fluree/db/query/exec/where.cljc (hunk 4 above)
[the remark this answers is not in the document]
Comment (cap10morgan) [reply src/fluree/db/query/exec/where.cljc:221]: Ohhhhh I just realized what you were referring to. Yeah, I went back and forth a little on this. In the end I decided I liked it better when the fuel-tracker code created and controlled the channel. But I could imagine a different version where the error channel gets passed into the fuel-tracker instead.

[the remark this answers is not in the document]
Comment (zonotope) [reply src/fluree/db/query/exec/where.cljc:221]: I see what you mean, and I can go either way as well. But I do like the idea of having one channel for errors that anything in the system can write to if there's a problem. That way you only need to listen to one spot without any relay code. But, that would probably need a refactor to implement anyway.

**Comment by zonotope** [comment]
> @dpetran & @zonotope let me know what you think of [e170b70](https://github.com/fluree/db/commit/e170b706b02f64b7e9be84d065084441b554f3a2)

👍🏾

File: src/fluree/db/util/core.cljc [hunk 5]
```diff
@@ -426,3 +426,12 @@
   (-> json-ld
       (get-first k)
       :id))
+
+(defn parse-opts
```
Comment (dpetran) [inline src/fluree/db/util/core.cljc:430]: I'm not sure I love it being in `util.core`, but I think it works. As we do the new transaction semantics I think this sort of thing will be consolidated more formally anyways.

Comment (cap10morgan) [reply src/fluree/db/util/core.cljc:430]: Do you see a better place for it? I didn't spot a better place for something shared by both queries and transactions.

**Comment by dpetran** [comment]
> @dpetran & @zonotope let me know what you think of [e170b70](https://github.com/fluree/db/commit/e170b706b02f64b7e9be84d065084441b554f3a2)

:+1:

**Pull request merged**